balancer-js/src/modules/swaps/joinExit/joinAndExit.ts

```diff
@@ -0,0 +1,309 @@
+  user: string,
+  relayerAddress: string,
+  wrappedNativeAsset: string,
+  slippage: string,
+  authorisation: string | undefined
+): {
+  to: string;
+  data: string;
+  rawCalls: string[];
+  inputs: (EncodeBatchSwapInput | ExitPoolData | EncodeJoinPoolInput)[];
+} {
+  // For each 'swap' create a swap/join/exit action
+  const actions = getActions(
+    swapInfo.tokenIn,
+    swapInfo.tokenOut,
+    swapInfo.swaps,
+    swapInfo.tokenAddresses,
+    slippage,
+    pools,
+    user,
+    relayerAddress
+  );
+  // Arrange action into order that will create minimal amount of calls
+  const orderedActions = orderActions(actions);
+
+  const calls: string[] = [];
+  const inputs: (EncodeBatchSwapInput | ExitPoolData | EncodeJoinPoolInput)[] =
+    [];
+  if (authorisation)
+    // Uses relayer to approve itself to act in behalf of the user
+    calls.push(
+      Relayer.encodeSetRelayerApproval(relayerAddress, true, authorisation)
+    );
+
+  // Create encoded call for each action
+  for (const action of orderedActions) {
+    if (action.type === ActionType.Exit || action.type === ActionType.Join) {
+      const pool = pools.find((p) => p.id === action.poolId);
+      if (pool === undefined)
+        throw new BalancerError(BalancerErrorCode.NO_POOL_DATA);
+      const { params, encoded } = action.callData(pool, wrappedNativeAsset);
+      calls.push(encoded as string);
+      inputs.push(params);
+    }
+    if (action.type === ActionType.BatchSwap) {
+      const { params, encoded } = action.callData();
+      calls.push(...encoded);
+      inputs.push(params);
+    }
+  }
+
+  // Safety check to make sure amounts/limits from calls match expected
+  checkAmounts(
+    orderedActions.map((a) => BigNumber.from(a.getAmountIn())),
+    orderedActions.map((a) => BigNumber.from(a.getAmountOut())),
+    swapInfo,
+    slippage
+  );
+  // encode relayer multicall
+  const callData = balancerRelayerInterface.encodeFunctionData('multicall', [
+    calls,
+  ]);
+
+  return {
+    to: relayerAddress,
+    data: callData,
+    rawCalls: calls,
+    inputs,
+  };
+}
+
+function checkAmounts(
+  amountsIn: BigNumber[],
+  amountsOut: BigNumber[],
+  swapInfo: SwapInfo,
+  slippage: string
+): void {
+  const totalIn = amountsIn.reduce(
+    (total = BigNumber.from(0), amount) => (total = total.add(amount))
+  );
+  const totalOut = amountsOut.reduce(
+    (total = BigNumber.from(0), amount) => (total = total.add(amount))
+  );
+  // totalIn should equal the original input swap amount
+  // totalOut should equal the return amount from SOR minus any slippage allowance
+  // console.log(totalIn.toString(), 'totalIn');
+  // console.log(swapInfo.swapAmount.toString(), 'swapInfo.swapAmount');
+  // console.log(totalOut.toString(), 'totalOut');
+  // console.log(
+  //   subSlippage(swapInfo.returnAmount, BigNumber.from(slippage)).toString(),
+  //   'slippage'
+  // );
+  // console.log(swapInfo.returnAmount.toString(), 'swapInfo.returnAmount');
```

Comment:
How about updating these to behave in the same way debugLog from generalisedExit does?